### PR TITLE
preserve param values case in case insensitive mode

### PR DESCRIPTION
**Pull request by misterdjules**
The implementation is a bit crude and simplistic, and the style is not great, but I thought it would be a good start to discuss and maybe iterate on.

Also note that this PR removes leftover `console.log` statements from the previous PR, my apologies for that.

### CODE REVIEW

**Comment by mcollina** [comment]
Can you run https://github.com/delvedor/find-my-way/blob/master/bench.js before and after on Node 8? It _shouldn't_ impact.

**Comment by mcollina** [comment]
> Also note that this PR removes leftover console.log statements from the previous PR, my apologies for that.

🎉 thanks for that!

**Comment by misterdjules** [comment]
Benchmark results:

Node version:

```
$ node --version
v8.12.0
$ 
```
Latest master branch:
```
$ node bench.js 
lookup static route x 30,722,064 ops/sec ±0.65% (90 runs sampled)
lookup dynamic route x 2,295,795 ops/sec ±0.94% (90 runs sampled)
lookup dynamic multi-parametric route x 1,257,638 ops/sec ±0.97% (89 runs sampled)
lookup dynamic multi-parametric route with regex x 1,134,217 ops/sec ±0.45% (95 runs sampled)
lookup long static route x 2,050,955 ops/sec ±0.54% (94 runs sampled)
lookup long dynamic route x 1,467,794 ops/sec ±0.55% (89 runs sampled)
lookup static versioned route x 6,081,052 ops/sec ±0.63% (89 runs sampled)
find static route x 33,708,958 ops/sec ±0.55% (90 runs sampled)
find dynamic route x 3,147,063 ops/sec ±0.88% (93 runs sampled)
find dynamic multi-parametric route x 1,638,160 ops/sec ±0.47% (92 runs sampled)
find dynamic multi-parametric route with regex x 1,287,578 ops/sec ±2.05% (89 runs sampled)
find long static route x 3,562,587 ops/sec ±1.04% (91 runs sampled)
find long dynamic route x 2,497,485 ops/sec ±0.90% (93 runs sampled)
find static versioned route x 7,352,454 ops/sec ±1.06% (92 runs sampled)
$
```
This PR:
```
$ node bench.js 
lookup static route x 27,923,275 ops/sec ±2.98% (87 runs sampled)
lookup dynamic route x 1,915,313 ops/sec ±3.27% (83 runs sampled)
lookup dynamic multi-parametric route x 1,165,372 ops/sec ±2.26% (87 runs sampled)
lookup dynamic multi-parametric route with regex x 1,072,591 ops/sec ±2.26% (89 runs sampled)
lookup long static route x 1,957,951 ops/sec ±1.12% (90 runs sampled)
lookup long dynamic route x 1,388,719 ops/sec ±0.80% (91 runs sampled)
lookup static versioned route x 6,034,244 ops/sec ±0.59% (90 runs sampled)
find static route x 31,514,577 ops/sec ±1.12% (89 runs sampled)
find dynamic route x 2,960,212 ops/sec ±1.63% (89 runs sampled)
find dynamic multi-parametric route x 1,557,106 ops/sec ±0.57% (92 runs sampled)
find dynamic multi-parametric route with regex x 1,285,181 ops/sec ±0.84% (95 runs sampled)
find long static route x 3,256,538 ops/sec ±2.77% (84 runs sampled)
find long dynamic route x 2,374,045 ops/sec ±0.52% (92 runs sampled)
find static versioned route x 6,801,250 ops/sec ±1.25% (91 runs sampled)
$ 
```

So it seems this PR introduces a ~5%-~10% performance regression? That doesn't seem great. Unless I'm missing something, I can look into that.

**Comment by misterdjules** [comment]
Benchmark ran on latest commit of this PR:

```
lookup static route x 28,508,842 ops/sec ±1.80% (85 runs sampled)
lookup dynamic route x 2,160,411 ops/sec ±1.69% (86 runs sampled)
lookup dynamic multi-parametric route x 1,258,515 ops/sec ±0.66% (93 runs sampled)
lookup dynamic multi-parametric route with regex x 1,088,146 ops/sec ±1.23% (89 runs sampled)
lookup long static route x 2,030,623 ops/sec ±1.04% (92 runs sampled)
lookup long dynamic route x 1,410,054 ops/sec ±1.35% (90 runs sampled)
lookup static versioned route x 5,929,448 ops/sec ±1.18% (92 runs sampled)
find static route x 33,738,252 ops/sec ±1.10% (92 runs sampled)
find dynamic route x 3,063,341 ops/sec ±1.39% (87 runs sampled)
find dynamic multi-parametric route x 1,559,624 ops/sec ±1.27% (89 runs sampled)
find dynamic multi-parametric route with regex x 1,256,616 ops/sec ±2.36% (85 runs sampled)
find long static route x 3,381,202 ops/sec ±1.15% (87 runs sampled)
find long dynamic route x 2,334,798 ops/sec ±0.99% (89 runs sampled)
find static versioned route x 6,832,293 ops/sec ±1.77% (88 runs sampled)
```

It seems slightly better but it still is a regression. 

We could also explore introducing an option to the `lookup()` function that would determine whether we want to pay the cost for having parameter values' not normalized (lower cased), thoughts?

File: node.js [hunk 1]
```diff
@@ -99,10 +99,14 @@ Node.prototype.findByLabel = function (path) {
   return this.children[path[0]]
 }
 
-Node.prototype.findChild = function (path, method) {
-  var child = this.children[path[0]]
+Node.prototype.findChild = function (path, method, caseSensitive) {
```
Comment (mcollina) [inline node.js]: I think we can pass the `caseSensitive` option when creating a node. Ideally Turbofan would see it's not changing and optimize for that.

Another option might be to do some prototype magic, and have a `NodeCaseInsensitive` constructor.

File: node.js [hunk 2]
```diff
@@ -99,10 +99,14 @@ Node.prototype.findByLabel = function (path) {
...
+  var normalizedPath = path
+  if (!caseSensitive) {
```
Comment (mcollina) [inline node.js]: In turbofan `this.caseSensitive === false` is faster.

**Comment by delvedor** [comment]
Closed in #115.

**Pull request closed without merging**